cmd/tools/rest/client.go

```diff
@@ -0,0 +1,215 @@
+// Copyright NetApp Inc, 2021 All rights reserved
+
+package rest
+
+import (
+	"bytes"
+	"crypto/tls"
+	"fmt"
+	"goharvest2/pkg/conf"
+	"goharvest2/pkg/errors"
+	"goharvest2/pkg/logging"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+)
+
+const (
+	// DefaultTimeout should be > than ONTAP's default REST timeout, which is 15 seconds for GET requests
+	DefaultTimeout = 30 * time.Second
+)
+
+type Client struct {
+	client   *http.Client
+	request  *http.Request
+	buffer   *bytes.Buffer
+	Logger   *logging.Logger
+	baseURL  string
+	password string
+	username string
+}
+
+func New(poller *conf.Poller) (*Client, error) {
+	var (
+		client         Client
+		httpclient     *http.Client
+		transport      *http.Transport
+		cert           tls.Certificate
+		timeout        time.Duration
+		addr           *string
+		url            string
+		useInsecureTLS bool
+		err            error
+	)
+
+	client = Client{}
+	client.Logger = logging.SubLogger("REST", "Client")
+
+	if addr = poller.Addr; addr == nil {
+		return nil, errors.New(errors.MISSING_PARAM, "addr")
+	}
+
+	if poller.IsKfs != nil && *poller.IsKfs {
+		url = "https://" + *addr + ":8443/"
+	} else {
+		url = "https://" + *addr + "/"
+	}
+	client.baseURL = url
+
+	// by default, enforce secure TLS, if not requested otherwise by user
+	if x := poller.UseInsecureTls; x != nil {
+		useInsecureTLS = *poller.UseInsecureTls
+	} else {
+		useInsecureTLS = false
+	}
+
+	// set authentication method
+	if poller.AuthStyle != nil && *poller.AuthStyle == "certificate_auth" {
+		certPath := value(poller.SslCert, "")
+		keyPath := value(poller.SslKey, "")
+		if certPath == "" {
+			return nil, errors.New(errors.MISSING_PARAM, "ssl_cert")
+		} else if keyPath == "" {
+			return nil, errors.New(errors.MISSING_PARAM, "ssl_key")
+		} else if cert, err = tls.LoadX509KeyPair(certPath, keyPath); err != nil {
+			return nil, err
+		}
+
+		transport = &http.Transport{
+			Proxy: http.ProxyFromEnvironment,
+			TLSClientConfig: &tls.Config{
+				Certificates:       []tls.Certificate{cert},
+				InsecureSkipVerify: useInsecureTLS},
```

Comment:
the code above is temporary. it will be removed once the Harvest rest client code is finished. The Harvest rest and zapi code should share the common parts like above